core/src/cost_tracker.rs

```diff
@@ -0,0 +1,356 @@
+//! `cost_tracker` keeps tracking tranasction cost per chained accounts as well as for entire block
+//! The main entry function is 'try_add', if success, it returns new block cost.
+//!
+use crate::cost_model::TransactionCost;
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+use std::collections::HashMap;
+
+#[derive(Debug, Clone)]
+pub struct CostTracker {
+    account_cost_limit: u32,
+    block_cost_limit: u32,
+    current_bank_slot: Slot,
+    cost_by_writable_accounts: HashMap<Pubkey, u32>,
+    block_cost: u32,
+}
+
+impl CostTracker {
+    pub fn new(chain_max: u32, package_max: u32) -> Self {
+        assert!(chain_max <= package_max);
+        Self {
+            account_cost_limit: chain_max,
+            block_cost_limit: package_max,
+            current_bank_slot: 0,
+            cost_by_writable_accounts: HashMap::new(),
+            block_cost: 0,
+        }
+    }
+
+    pub fn reset_if_new_bank(&mut self, slot: Slot) {
+        if slot != self.current_bank_slot {
+            self.current_bank_slot = slot;
+            self.cost_by_writable_accounts.clear();
+            self.block_cost = 0;
+        }
+    }
+
+    pub fn try_add(&mut self, transaction_cost: TransactionCost) -> Result<u32, &'static str> {
+        let cost = transaction_cost.account_access_cost + transaction_cost.execution_cost;
+        self.would_fit(&transaction_cost.writable_accounts, &cost)?;
+
+        self.add_transaction(&transaction_cost.writable_accounts, &cost);
+        Ok(self.block_cost)
+    }
+
+    fn would_fit(&self, keys: &[Pubkey], cost: &u32) -> Result<bool, &'static str> {
+        // check against the total package cost
+        if self.block_cost + cost > self.block_cost_limit {
+            return Err("would exceed block cost limit");
+        }
+
+        // chech if the transaction itself is more costly than the account_cost_limit
+        if *cost > self.account_cost_limit {
+            return Err("Transaction is too expansive, exceeds account cost limit");
+        }
+
+        // check each account against account_cost_limit,
+        for account_key in keys.iter() {
+            match self.cost_by_writable_accounts.get(&account_key) {
+                Some(chained_cost) => {
+                    if chained_cost + cost > self.account_cost_limit {
+                        return Err("would exceed account cost limit");
+                    } else {
+                        continue;
+                    }
+                }
+                None => continue,
+            }
+        }
+
+        Ok(true)
```

Comment:
Does anything use this value? If not, just use the empty return `()`

Reply:
good call! changed return type to `Result<(), str>` and returning `Ok(())`